README.md
# SpeechBubble

A flutter widget that emulates a speech bubble.

This widget is modeled after the red Instagram notification popups.

![Screenshot](screenshots/igicons.jpg)

<div class="center">
    <img width="359" height="693" src="screenshots/example2.jpg">
  </div>

# Installing

Dart Pub: https://pub.dartlang.org/packages/speech_bubble#-installing-tab-

1. Depend on it

   Add this to your package's pubspec.yaml file:

```dependencies:
  speech_bubble: ^0.0.5
```

2. Install it

   You can install packages from the command line:

```
$ flutter packages get
```

Alternatively, your editor might support flutter packages get. Check the docs for your editor to learn more.

3. Import it

   Now in your Dart code, you can use:

```
import 'package:speech_bubble/speech_bubble.dart';
```

# Usage

Creates a widget that emulates a speech bubble.
Could be used for a tooltip, or as a pop-up notification, etc.

<div class="center">
    <img width="359" height="693" src="screenshots/example1.jpg">
  </div>

# Properties

- child -> Widget

The child contained by the SpeechBubble.

- nipLocation -> NipLocation

The location of the nip of the speech bubble.
Use the NipLocation enum, either TOP, RIGHT, BOTTOM, or LEFT.
The nip will automatically center to the side that it is assigned.

- color -> Color

The color of the body of the SpeechBubble and nip.
Defaultly red.

- borderRadius -> double

The borderRadius of the SpeechBubble.
The SpeechBubble is built with a circular border radius on all 4 corners.

- height -> double

The explicitly defined height of the SpeechBubble.
The SpeechBubble will defaultly enclose its child.

- width -> double

The explicitly defined width of the SpeechBubble.
The SpeechBubble will defaultly enclose its child.

- padding -> Widget

The padding widget between the child and the edges of the SpeechBubble.

# Example Code

```dart
import 'package:flutter/material.dart';
import 'package:speech_bubble/speech_bubble.dart';

void main() => runApp(new MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => new _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      home: new Scaffold(
        appBar: new AppBar(
          title: const Text('Plugin example app'),
        ),
        body: new Center(
          child: new SpeechBubble(
            nipLocation: NipLocation.BOTTOM,
            // child: Column(
            //   mainAxisSize: MainAxisSize.min,
            //   children: <Widget>[
            //     Text("Give your users some guided instruction"),
            //     Text("From the inside of a Speech Bubble")
            //   ],
            // ),
            child: Row(
              mainAxisSize: MainAxisSize.min,
              children: <Widget>[
                Icon(
                  Icons.favorite,
                  color: Colors.white,
                ),
                Padding(
                  padding: const EdgeInsets.all(4.0),
                ),
                Text(
                  "1",
                  style: TextStyle(
                    color: Colors.white,
                    fontSize: 18.0,
                  ),
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}
```


CHANGELOG.md
## 0.0.10 - 09-16-2021

- Fix "use a deprecated version of the Android embedding" converting this plugin to a package

## 0.0.9 - 04-10-2021

- Up minimum flutter sdk

## 0.0.8 - 04-10-2021

- Added null safety

## 0.0.6 - 9-18-2019

- Changed type of padding

## 0.0.5 - 8-13-2018

- Updated readme

## 0.0.4 - 8-13-2018

- Updated change log

## 0.0.3 - 8-13-2018

- Fixed fatal gradle error

## 0.0.2 - 8-13-2018

- Updated readme

## 0.0.1 - 8-11-2018

- Initial release.


lib/speech_bubble.dart
import 'dart:math';

import 'package:flutter/material.dart';

enum NipLocation {
  TOP,
  RIGHT,
  BOTTOM,
  LEFT,
  BOTTOM_RIGHT,
  BOTTOM_LEFT,
  TOP_RIGHT,
  TOP_LEFT
}

const defaultNipHeight = 10.0;

class SpeechBubble extends StatelessWidget {
  /// Creates a widget that emulates a speech bubble.
  /// Could be used for a tooltip, or as a pop-up notification, etc.
  SpeechBubble(
      {Key? key,
      required this.child,
      this.nipLocation = NipLocation.BOTTOM,
      this.color = Colors.redAccent,
      this.borderRadius = 4.0,
      this.height,
      this.width,
      this.padding,
      this.nipHeight = defaultNipHeight,
      this.offset = Offset.zero})
      : super(key: key);

  /// The [child] contained by the [SpeechBubble]
  final Widget child;

  /// The location of the nip of the speech bubble.
  ///
  /// Use [NipLocation] enum, either [TOP], [RIGHT], [BOTTOM], or [LEFT].
  /// The nip will automatically center to the side that it is assigned.
  final NipLocation nipLocation;

  /// The color of the body of the [SpeechBubble] and nip.
  /// Defaultly red.
  final Color color;

  /// The [borderRadius] of the [SpeechBubble].
  /// The [SpeechBubble] is built with a
  /// circular border radius on all 4 corners.
  final double borderRadius;

  /// The explicitly defined height of the [SpeechBubble].
  /// The [SpeechBubble] will defaultly enclose its [child].
  final double? height;

  /// The explicitly defined width of the [SpeechBubble].
  /// The [SpeechBubble] will defaultly enclose its [child].
  final double? width;

  /// The padding between the child and the edges of the [SpeechBubble].
  final EdgeInsetsGeometry? padding;

  /// The nip height
  final double nipHeight;

  final Offset offset;

  Widget build(BuildContext context) {
    Offset? nipOffset;
    AlignmentGeometry? alignment;
    var rotatedNipHalfHeight = getNipHeight(nipHeight) / 2;
    var offset = nipHeight / 2 + rotatedNipHalfHeight;
    switch (nipLocation) {
      case NipLocation.TOP:
        nipOffset = Offset(0.0, -offset + rotatedNipHalfHeight);
        alignment = Alignment.topCenter;
        break;
      case NipLocation.RIGHT:
        nipOffset = Offset(offset - rotatedNipHalfHeight, 0.0);
        alignment = Alignment.centerRight;
        break;
      case NipLocation.BOTTOM:
        nipOffset = Offset(0.0, offset - rotatedNipHalfHeight);
        alignment = Alignment.bottomCenter;
        break;
      case NipLocation.LEFT:
        nipOffset = Offset(-offset + rotatedNipHalfHeight, 0.0);
        alignment = Alignment.centerLeft;
        break;
      case NipLocation.BOTTOM_LEFT:
        nipOffset = this.offset +
            Offset(
                offset - rotatedNipHalfHeight, offset - rotatedNipHalfHeight);
        alignment = Alignment.bottomLeft;
        break;
      case NipLocation.BOTTOM_RIGHT:
        nipOffset = this.offset +
            Offset(
                -offset + rotatedNipHalfHeight, offset - rotatedNipHalfHeight);
        alignment = Alignment.bottomRight;
        break;
      case NipLocation.TOP_LEFT:
        nipOffset = this.offset +
            Offset(
                offset - rotatedNipHalfHeight, -offset + rotatedNipHalfHeight);
        alignment = Alignment.topLeft;
        break;
      case NipLocation.TOP_RIGHT:
        nipOffset = this.offset +
            Offset(
                -offset + rotatedNipHalfHeight, -offset + rotatedNipHalfHeight);
        alignment = Alignment.topRight;
        break;
      default:
    }

    return Stack(
      alignment: alignment!,
      children: <Widget>[
        speechBubble(),
        nip(nipOffset!),
      ],
    );
  }

  Widget speechBubble() {
    return Material(
      borderRadius: BorderRadius.all(
        Radius.circular(borderRadius),
      ),
      color: color,
      elevation: 1.0,
      child: Container(
        height: height,
        width: width,
        padding: padding ?? const EdgeInsets.all(8.0),
        child: child,
      ),
    );
  }

  Widget nip(Offset nipOffset) {
    return Transform.translate(
      offset: nipOffset,
      child: RotationTransition(
        turns: AlwaysStoppedAnimation(45 / 360),
        child: Material(
          borderRadius: BorderRadius.all(
            Radius.circular(1.5),
          ),
          color: color,
          child: Container(
            height: nipHeight,
            width: nipHeight,
          ),
        ),
      ),
    );
  }

  double getNipHeight(double nipHeight) => sqrt(2 * pow(nipHeight, 2));
}


android/src/main/java/com/plushundred/speechbubble/SpeechBubblePlugin.java
package com.plushundred.speechbubble;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

/** SpeechBubblePlugin */
public class SpeechBubblePlugin implements MethodCallHandler {
  /** Plugin registration. */
  public static void registerWith(Registrar registrar) {
    final MethodChannel channel = new MethodChannel(registrar.messenger(), "speech_bubble");
    channel.setMethodCallHandler(new SpeechBubblePlugin());
  }

  @Override
  public void onMethodCall(MethodCall call, Result result) {
    if (call.method.equals("getPlatformVersion")) {
      result.success("Android " + android.os.Build.VERSION.RELEASE);
    } else {
      result.notImplemented();
    }
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:speech_bubble/speech_bubble.dart';

void main() => runApp(new MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => new _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      home: new Scaffold(
        appBar: new AppBar(
          title: const Text('Plugin example app'),
        ),
        body: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Center(
              child: new SpeechBubble(
                nipLocation: NipLocation.BOTTOM,
                child: Row(
                  mainAxisSize: MainAxisSize.min,
                  children: <Widget>[
                    Icon(
                      Icons.favorite,
                      color: Colors.white,
                    ),
                    Padding(
                      padding: const EdgeInsets.all(4.0),
                    ),
                    Text(
                      "1",
                      style: TextStyle(
                        color: Colors.white,
                        fontSize: 18.0,
                      ),
                    ),
                  ],
                ),
              ),
            ),
            SizedBox(height: 20,),
            Center(
              child: new SpeechBubble(
                nipLocation: NipLocation.BOTTOM_LEFT,
                nipHeight: 20,
                child: Row(mainAxisSize: MainAxisSize.min, children: <Widget>[
                  Text(
                    "You can also change the nip size!",
                    style: TextStyle(
                      color: Colors.white,
                      fontSize: 18.0,
                    ),
                  )
                ]),
              ),
            )
          ],
        ),
      ),
    );
  }
}
